torch/csrc/api/include/torch/data/dataloader/stateless_impl.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <torch/data/dataloader/impl.h>
+#include <torch/data/worker_exception.h>
+
+#include <torch/csrc/utils/memory.h>
+
+#include <c10/util/Exception.h>
+
+#include <cstddef>
+#include <thread>
+#include <utility>
+
+namespace torch {
+namespace data {
+namespace detail {
+namespace dataloader {
+template <typename Dataset, typename Sampler>
+class StatelessImpl : public Impl<
+                          Dataset,
+                          typename Dataset::BatchType,
+                          typename Sampler::BatchRequestType> {
+ public:
+  using super = Impl<
+      Dataset,
+      typename Dataset::BatchType,
+      typename Sampler::BatchRequestType>;
+  using super::options_;
+  using super::pop_result;
+  using super::prefetch;
+  using super::push_job;
+  using super::workers_;
+  using typename super::BatchRequestType;
+  using typename super::BatchType;
+  using typename super::Result;
+
+  StatelessImpl(Dataset dataset, DataLoaderOptions options, Sampler sampler)
+      : super(std::move(options)), sampler_(std::move(sampler)) {
+    for (size_t w = 0; w < options_.workers; ++w) {
+      // Here we copy the dataset into the worker thread closure. Each worker
+      // has its own copy of the dataset. This means the dataset must be
+      // trivially copiable, or else we don't expect more than one worker to
+      // be in use.
+      workers_.emplace_back(
+          [this, dataset]() mutable { this->worker_thread(dataset); });
+    }
+    if (options_.workers == 0) {
+      main_thread_dataset_ = torch::make_unique<Dataset>(std::move(dataset));
+    }
+  }
+
+  void reset(bool prefetch = true) override {
+    sampler_.reset();
+    super::reset(prefetch);
+  }
+
+ private:
+  optional<BatchType> next() {
+    optional<BatchType> batch;
+    if (options_.workers > 0) {
+      optional<Result> result = pop_result();
+      if (result) {
+        if (result->exception) {
+          throw WorkerException(result->exception);
+        } else {
+          AT_ASSERT(result->batch.has_value());
+          batch = std::move(result->batch);
+          prefetch(1);
```

Comment:
I this this is pretty much the same code as in the stateful case, except there we can get rid of `batch` by simply returning from here.

Reply:
I'll see if I can consolidate more